[SchemaRegistry] generate sr from tsp

swathipil opened the pull request:
Generating from tsp here: https://github.com/Azure/azure-rest-api-specs/pull/25168

TODO Re-generate after:
- [x] Multiple content types support has been added: https://github.com/Azure/autorest.python/issues/1526 OR add unimplemented operations to patch file
     - As a workaround for testing purposes, currently modifying local typespec file to only take a single content type.
     - Following issue should be fixed: in models/_enums.py, content type enum is generated like this when there are multiple content types (https://github.com/Azure/autorest.python/issues/2082):
```
    APPLICATION_JSON;_SERIALIZATION=_AVRO = "application/json; serialization=Avro"
    APPLICATION_JSON;_SERIALIZATION=JSON = "application/json; serialization=json"
    TEXT_PLAIN;_CHARSET=UTF_8 = "text/plain; charset=utf-8"
```
- [x] In certain operations, schema_content is being passed through w/ `json.dumps(content, cls=AzureJSONEncoder)`, but this should be removed. Should be `_content = content`. ~Fix is currently being worked on.~ Fix is in.
- [ ] NON-BLOCKING: The "Accept" request header isn't the same as before on all operations. Check with Izzy on default types for Accept. (issue here: https://github.com/Azure/autorest.python/issues/2122)
- [x] NON-BLOCKING: Need extras_require in setup.py, so don't overwrite --> [basic-setup-py=False in tspconfig doesn't work](https://github.com/Azure/autorest.python/blob/2f14606c12e68c3c6172456ffba15706407f6c8a/packages/typespec-python/src/lib.ts#L4) (issue: https://github.com/Azure/autorest.python/issues/2119)
  - not committing the generated setup.py for now
- [x] BLOCKING: generated code has [mypy/pylint errors](https://dev.azure.com/azure-sdk/public/_build/results?buildId=3117682&view=logs&jobId=69247760-d00b-59ba-802c-4e283c193199&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=e3f8f308-6e5f-58aa-3a2a-fe425958de32) (issue here: https://github.com/Azure/autorest.python/issues/2148)
  - manually fixing to unblock
- [x] BLOCKING: Make models namespace private. --> update `@@access` to private for all models in client.tsp
- [x] BLOCKING: SchemaProperties.format should return SchemaFormat, not string (issue here: https://github.com/Azure/autorest.python/issues/2276)  --> Added SchemaProperties/SchemaFormat to the _patch.py file, so we're no longer generating it.
- [ ] BLOCKING: https://github.com/microsoft/typespec/issues/2853 --> address in separate PR
- [x] BLOCKING: tsp bug - https://github.com/Azure/autorest.python/issues/2671 --> fixing manually
- [x] BLOCKING: tsp bug - https://github.com/Azure/autorest.python/issues/2670 --> fixing manually

msyyc commented:
@swathipil I can't reproduce the error. According to log info, it may be related with some unexpected file. So please clear your local folder. After that, merge latest main branch and try again. If you still meet the error, please share the detailed log.

swathipil commented:
Hi @msyyc - I just cleared my local cache, re-synced to [[this latest tsp commit](https://github.com/Azure/azure-rest-api-specs/pull/25168/commits/a1525d57058dcba9eea8a7c9318f536785272a12)] using
`azure-schemaregistry>..\..\..\eng\common\scripts\TypeSpec-Project-Sync.ps1 .` and generated using `azure-schemaregistry>..\..\..\eng\common\scripts\TypeSpec-Project-Generate.ps1 .`.  Here is the full error log:

```
Using module powershell-yaml with version 0.4.7.
Reading configuration from C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\tsp-location.yaml
Generating from C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry
Copying package.json from C:\Users\swathip\repos\azure-sdk-for-python\eng\common\scripts\..\..\emitter-package.json

added 102 packages, and audited 103 packages in 54s

9 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npx tsp compile C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\client.tsp --emit @azure-tools/typespec-python --option @azure-tools/typespec-python.emitter-output-dir=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/
TypeSpec compiler v0.47.0

Not going to generate operation "build_schema_registry_get_schema_id_by_content_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "build_schema_registry_register_schema_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "get_schema_id_by_content" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "register_schema" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 69, in <module>
    BlackScriptPlugin(output_folder=args.output_folder, **unknown_args).process()
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 34, in process
    list(
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 48, in format_file
    file_content = self.read_file(file)
                   ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\__init__.py", line 47, in read_file
    return fd.read()
           ^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
  File "C:\Users\swathip\AppData\Local\Programs\Python\Python311\Lib\encodings\utf_8_sig.py", line 69, in _buffer_decode
    return codecs.utf_8_decode(input, errors, final)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf6 in position 122: invalid start byte
Error: Command failed: py -3 C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run_cadl.py --output-folder=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/ --cadl-file=C:/Users/swathip/AppData/Local/Temp/cadl-codegen/typespec-python-yaml-map7f7078c3-4e81-4611-b31a-71f8ebd3d2a7.yaml --basic-setup-py=true --package-version=1.3.0b3 --package-dir=azure-schemaregistry --package-name=azure-schemaregistry --package-mode=dataplane
ExternalError: Emitter "@azure-tools/typespec-python" failed!
File issue at https://github.com/Azure/autorest.python/issues

Error: Command failed: C:\Program Files\nodejs\node.exe C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run-python3.js C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run_cadl.py --output-folder=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/ --cadl-file=C:/Users/swathip/AppData/Local/Temp/cadl-codegen/typespec-python-yaml-map7f7078c3-4e81-4611-b31a-71f8ebd3d2a7.yaml --basic-setup-py=true --package-version=1.3.0b3 --package-dir=azure-schemaregistry --package-name=azure-schemaregistry --package-mode=dataplane
Not going to generate operation "build_schema_registry_get_schema_id_by_content_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "build_schema_registry_register_schema_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "get_schema_id_by_content" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "register_schema" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 69, in <module>
    BlackScriptPlugin(output_folder=args.output_folder, **unknown_args).process()
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 34, in process
    list(
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 48, in format_file
    file_content = self.read_file(file)
                   ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\__init__.py", line 47, in read_file
    return fd.read()
           ^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
  File "C:\Users\swathip\AppData\Local\Programs\Python\Python311\Lib\encodings\utf_8_sig.py", line 69, in _buffer_decode
    return codecs.utf_8_decode(input, errors, final)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf6 in position 122: invalid start byte
Error: Command failed: py -3 C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run_cadl.py --output-folder=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/ --cadl-file=C:/Users/swathip/AppData/Local/Temp/cadl-codegen/typespec-python-yaml-map7f7078c3-4e81-4611-b31a-71f8ebd3d2a7.yaml --basic-setup-py=true --package-version=1.3.0b3 --package-dir=azure-schemaregistry --package-name=azure-schemaregistry --package-mode=dataplane

    at checkExecSyncError (node:child_process:885:11)
    at execFileSync (node:child_process:921:15)
    at Object.$onEmit [as emitFunction] (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@azure-tools/typespec-python/dist/src/emitter.js:42:9)
    at async runEmitter (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:563:13)
    at async compile (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:257:9)
    at async compileOnce (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/cli/actions/compile/compile.js:37:25)
    at async compileAction (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/cli/actions/compile/compile.js:21:9)
    at runEmitter (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:583:19)
    at async compile (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:257:9)
    at async compileOnce (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/cli/actions/compile/compile.js:37:25)
    at async compileAction (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/cli/actions/compile/compile.js:21:9)
```

I know that the `unable to generate this operation` errors are because there are multiple content types in the request for those ops. 

However, I don't think the `utf_8_decode` errors have to do with that. Even when I update the content type to only a single type in the tsp file, then re-generate, I still get the `utf_8_decode` errors. Thanks!

msyyc commented:
I advise you checkout to another branch and create new PR. Here is cmd:
```
D:\dev4\azure-sdk-for-python> git remote add Azure https://github.com/Azure/azure-sdk-for-python.git (only need once)
D:\dev4\azure-sdk-for-python> git fetch Azure main
D:\dev4\azure-sdk-for-python> git checkout Azure/main
D:\dev4\azure-sdk-for-python> git checkout -b schemaregistry-2023-08-10
D:\dev4\azure-sdk-for-python> git checkout -b schemaregistry-2023-08-10
D:\dev4\azure-sdk-for-python>  pwsh eng\common\scripts\TypeSpec-Project-Process.ps1 https://github.com/Azure/azure-rest-api-specs/blob/a1525d57058dcba9eea8a7c9318f536785272a12/specification/schemaregistry/SchemaRegistry/tspconfig.yaml
```

I can succeed the up cmd:
![image](https://github.com/Azure/azure-sdk-for-python/assets/70930885/ab1c459d-30cf-4e04-8272-d4cd29b71f21)

Here is my PR: https://github.com/Azure/azure-sdk-for-python/pull/31558

swathipil commented:
Thanks @msyyc. It looks like the error is being raised from the Python 3.11 encodings lib. Which version of Python are you using in the environment where you run the command above?

msyyc commented:
> Thanks @msyyc. It looks like the error is being raised from the Python 3.11 encodings lib. Which version of Python are you using in the environment where you run the command above?

I am using python3.8, let me try python3.11

swathipil commented:
Hi @msyyc - I just tested what you mentioned in [this comment](https://github.com/Azure/azure-sdk-for-python/pull/31475#issuecomment-1672461864) (created a branch off of main then ran `Typespec-Project-Process.ps1`. This still gave me the following:
 ```
 azure-sdk-for-python> pwsh eng\common\scripts\TypeSpec-Project-Process.ps1 https://github.com/Azure/azure-rest-api-specs/blob/a1525d57058dcba9eea8a7c9318f536785272a12/specification/schemaregistry/SchemaRegistry/tspconfig.yaml
Using module powershell-yaml with version 0.4.7.
creating tsp-location.yaml in C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry
updated tsp-location.yaml commit to a1525d57058dcba9eea8a7c9318f536785272a12                 
updated tsp-location.yaml repo to Azure/azure-rest-api-specs                                 
updated tsp-location.yaml directory to specification/schemaregistry/SchemaRegistry           
updated tsp-location.yaml additionalDirectories to                                           
finished updating tsp-location.yaml in C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry
Calling TypeSpec-Project-Sync.ps1                                                            
Using module powershell-yaml with version 0.4.7.
Reading configuration from C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\tsp-location.yaml
Checking remote origin  https://github.com/swathipil/azure-sdk-for-python.git (fetch)        
Found git remote https://github.com/Azure/azure-rest-api-specs.git                           
from tsplocation.yaml 'repo' is: Azure/azure-rest-api-specs                                  
Setting up sparse clone for azure-schemaregistry at C:\Users\swathip\repos\sparse-spec\azure-schemaregistry
Updating sparse checkout file with directory:specification/schemaregistry/SchemaRegistry     
git checkout commit: a1525d57058dcba9eea8a7c9318f536785272a12
warning: unrecognized pattern: 'specification/schemaregistry/SchemaRegistry'              ]
warning: disabling cone pattern matching
warning: unrecognized pattern: 'specification/schemaregistry/SchemaRegistry'
warning: disabling cone pattern matching
Copying spec from C:\Users\swathip\repos\sparse-spec\azure-schemaregistry\specification\schemaregistry\SchemaRegistry to C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/TempTypeSpecFiles
Calling TypeSpec-Project-Generate.ps1                                                        
Using module powershell-yaml with version 0.4.7.
Reading configuration from C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\tsp-location.yaml
Generating from C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry
Copying package.json from C:\Users\swathip\repos\azure-sdk-for-python\eng\common\scripts\..\..\emitter-package.json
npx tsp compile C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\main.tsp --emit @azure-tools/typespec-python --option @azure-tools/typespec-python.emitter-output-dir=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/
Not going to generate operation "build_schema_registry_get_schema_id_by_content_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "build_schema_registry_register_schema_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize    
Not going to generate operation "get_schema_id_by_content" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "register_schema" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 69, in <module>
    BlackScriptPlugin(output_folder=args.output_folder, **unknown_args).process()
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 34, in process
    list(
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 48, in format_file
    file_content = self.read_file(file)
                   ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\__init__.py", line 47, in read_file
    return fd.read()
           ^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
  File "C:\Users\swathip\AppData\Local\Programs\Python\Python311\Lib\encodings\utf_8_sig.py", line 69, in _buffer_decode
    return codecs.utf_8_decode(input, errors, final)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf6 in position 122: invalid start byte 
Error: Command failed: py -3 C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run_cadl.py --output-folder=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/ --cadl-file=C:/Users/swathip/AppData/Local/Temp/cadl-codegen/typespec-python-yaml-map73ba96fe-191d-4507-b5fa-fe0ce702afe9.yaml --basic-setup-py=true --package-version=1.3.0b3 --package-dir=azure-schemaregistry --package-name=azure-schemaregistry --package-mode=dataplane
ExternalError: Emitter "@azure-tools/typespec-python" failed!
File issue at https://github.com/Azure/autorest.python/issues

Error: Command failed: C:\Program Files\nodejs\node.exe C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run-python3.js C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run_cadl.py --output-folder=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/ --cadl-file=C:/Users/swathip/AppData/Local/Temp/cadl-codegen/typespec-python-yaml-map73ba96fe-191d-4507-b5fa-fe0ce702afe9.yaml --basic-setup-py=true --package-version=1.3.0b3 --package-dir=azure-schemaregistry --package-name=azure-schemaregistry --package-mode=dataplane
Not going to generate operation "build_schema_registry_get_schema_id_by_content_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "build_schema_registry_register_schema_request" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize    
Not going to generate operation "get_schema_id_by_content" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Not going to generate operation "register_schema" because we are unable to generate this type of operation right now. Please write your own custom operation in the "_patch.py" file following https://aka.ms/azsdk/python/dpcodegen/python/customize
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 69, in <module>
    BlackScriptPlugin(output_folder=args.output_folder, **unknown_args).process()
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 34, in process
    list(
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\black\__init__.py", line 48, in format_file
    file_content = self.read_file(file)
                   ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\TempTypeSpecFiles\SchemaRegistry\node_modules\@autorest\python\autorest\__init__.py", line 47, in read_file
    return fd.read()
           ^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
  File "C:\Users\swathip\AppData\Local\Programs\Python\Python311\Lib\encodings\utf_8_sig.py", line 69, in _buffer_decode
    return codecs.utf_8_decode(input, errors, final)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf6 in position 122: invalid start byte 
Error: Command failed: py -3 C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@autorest/python/run_cadl.py --output-folder=C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry/ --cadl-file=C:/Users/swathip/AppData/Local/Temp/cadl-codegen/typespec-python-yaml-map73ba96fe-191d-4507-b5fa-fe0ce702afe9.yaml --basic-setup-py=true --package-version=1.3.0b3 --package-dir=azure-schemaregistry --package-name=azure-schemaregistry --package-mode=dataplane

    at checkExecSyncError (node:child_process:885:11)
    at execFileSync (node:child_process:921:15)
    at Object.$onEmit [as emitFunction] (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@azure-tools/typespec-python/dist/src/emitter.js:42:9)
    at async runEmitter (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:563:13)
    at async compile (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:257:9)
    at async compileOnce (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/cli/actions/compile/compile.js:37:25)
    at async compileAction (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/cli/actions/compile/compile.js:21:9)
    at runEmitter (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:583:19)
    at async compile (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/program.js:257:9)
    at async compileOnce (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dist/src/core/cli/actions/compile/compile.js:37:25)
    at async compileAction (file:///C:/Users/swathip/repos/azure-sdk-for-python/sdk/schemaregistry/azure-schemaregistry/TempTypeSpecFiles/SchemaRegistry/node_modules/@typespec/compiler/dWrite-Error: Failed to generate sdk project at C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry
```

In any case, I don't think it's blocking me so feel free to de-prioritize this. Also, please let me know if I should create an issue in the autorest repo so you can track this.

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_operations.py:
```diff
@@ -0,0 +1,746 @@
+        if _stream:
+            deserialized = response.iter_bytes()
+        else:
+            deserialized = _deserialize(
```
this raises the following error:
```  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\azure\schemaregistry\_model_base.py", line 684, in _deserialize_with_callable
    return typing.cast(typing.Callable[[typing.Any], typing.Any], deserializer)(value)    
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^    
  File "C:\Users\swathip\repos\azure-sdk-for-python\sdk\schemaregistry\azure-schemaregistry\azure\schemaregistry\_model_base.py", line 255, in _deserialize_bytes
    return bytes(base64.b64decode(attr))
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\swathip\AppData\Local\Programs\Python\Python311\Lib\base64.py", line 88, in b64decode
    return binascii.a2b_base64(s, strict_mode=validate)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
binascii.Error: Incorrect padding```

Wonder if this has to do with the utf8_decode bug?

swathipil replied:
needed to pass stream=True into client call

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_operations.py:
```diff
@@ -0,0 +1,746 @@
+            'cls', None
+        )
+
+        _content = json.dumps(schema_content, cls=AzureJSONEncoder)  # type: ignore
```
different from current operation behavior.

swathipil replied:
passing a string into this results in an escape `\` being added to all the quotations in the string, which changes the content.

swathipil replied:
Izzy's working on a PR to fix this

swathipil commented on sdk/schemaregistry/azure-schemaregistry/setup.py:
```diff
@@ -1,78 +1,70 @@
-# Change the PACKAGE_NAME only to change folder and different name
 PACKAGE_NAME = "azure-schemaregistry"
-PACKAGE_PPRINT_NAME = "Schema Registry"
+PACKAGE_PPRINT_NAME = "Azure Schemaregistry"
```
should be "Azure Schema Registry"

swathipil replied:
need to not overwrite setup.py

swathipil commented:
/azp run python - schemaregistry - ci

swathipil commented on sdk/schemaregistry/azure-schemaregistry/setup.py:
```diff
@@ -1,78 +1,70 @@
+        "azure-core<2.0.0,>=1.28.0",
+        "typing-extensions>=4.3.0; python_version<'3.8.0'",
     ],
-    extras_require={
```
add this back

swathipil replied:
need to not overwrite setup.py

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[azure-schemaregistry](https://apiview.dev/Assemblies/Review/79dbf3282c51418e8038af11a837431e)

l0lawrence commented on sdk/schemaregistry/azure-schemaregistry/setup.py:
```diff
@@ -13,7 +13,7 @@
 # Change the PACKAGE_NAME only to change folder and different name
 PACKAGE_NAME = "azure-schemaregistry"
-PACKAGE_PPRINT_NAME = "Schema Registry"
+PACKAGE_PPRINT_NAME = "Azure Schema Registry"
```
do you want this to change here? Do you need to change the naming in your tsp-location file?

swathipil replied:
Yes, we want this change. So I'm actually trying to figure out the best way to do this rn + responding to Yuchao. But if setup.py is not overwritten (which it may not be b/c we need the extras_require), we want to add this in. If it is overwritten by the generated code, this change will be made anyway.

l0lawrence commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,985 @@
+        **kwargs
+    )
+
+class GeneratedClient(ServiceClientGenerated):
```
Why a generated client here with custom methods instead of adding these to the OperationsMixin in _patch.py?

swathipil replied:
ahh, I def forgot that models/operations have their own patches. Will add this in there, thanks :)

l0lawrence commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,985 @@
+        """
+
+
+class MessageContent(TypedDict):
```
The type of this in ApiView changed to ForwardRef(bytes/str)

swathipil replied:
no that was unintentional, will fix!

swathipil replied:
so I copied this over exactly as it was. I think it could be that something changed with the apiview generator? will look into it...

swathipil replied:
Travis is not sure either, but says maybe it's the Python version update. (Not exactly sure why that would be causing this though.)

l0lawrence commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_patch.py:
```diff
@@ -0,0 +1,294 @@
+
+class SchemaRegistryClientOperationsMixin(GeneratedClientOperationsMixin):
+    """
+    Implement custom operations.
```
Are these custom operations here and not in the tsp because of the issue with multi-content type?

swathipil replied:
So these ops are in the tsp. However, they don't generate b/c the generation doesn't currently support multiple content types in the request.

swathipil replied:
so yeah, what you said lol

l0lawrence commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_patch.py:
```diff
@@ -0,0 +1,294 @@
+    """
+
+    @distributed_trace
+    def get_schema_id_by_content(  # pylint: disable=inconsistent-return-statements
```
I dont see this in the apiview under the client? Am I missing something?

l0lawrence replied:
ah nvm I see the client _patch

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,724 @@
+    """
+    SchemaRegistryClient is a client for registering and retrieving schemas from the Azure Schema Registry service.
+
+    :param str fully_qualified_namespace: The Schema Registry service fully qualified host name.
```
Is this different than the endpoint parameter? If not, I'm curious why we didnt use `@projectedName` in the typespec to generate this property name if that's the goal for Python.

catalinaperalta replied:
If this name is meant to be used by all generated sdks, then I would set the scope in `@projectedName` to `client`

swathipil replied:
Same as the endpoint parameter. I just copied over what was in the swagger/other service typespecs, b/c it seems like using "endpoint" is the convention there. And I think we decided on fully_qualified_namespace here to follow SDK convention I think. I didn't know projectedName was a thing I could use here 🙂 will add it to the tsp!

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,724 @@
+
+class SchemaRegistryClient(object):
+    """
+    SchemaRegistryClient is a client for registering and retrieving schemas from the Azure Schema Registry service.
+
+    :param str fully_qualified_namespace: The Schema Registry service fully qualified host name.
```
We should be able to add these improvements in the tsp

swathipil replied:
qq: when you say improvements here, do you mean using the projectedName?

catalinaperalta replied:
I was referring to the client description and the endpoint param rename

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,724 @@
+        ]
+
+
+class SchemaFormat(str, Enum, metaclass=CaseInsensitiveEnumMeta):
```
Curious why these models arent part of the generated code?

swathipil replied:
hmm, no reason other than I didn't know we could/wanted to do that, since we were previously adding these models manually. But if we could generate them, that would be great!

swathipil replied:
actually, one reason to not generate this SchemaFormat model specifically is b/c different languages use different casing for the enum values here. "Json" vs. "JSON". I think technically, all the clients should be doing case insensitive checks here anyway, so maybe we still could generate it? 🤔

catalinaperalta replied:
@swathipil and I synced on this offline. It's better to add the models into the tsp and do this migration once. If some of the SDKs made larger changes to the service model schemas then they can use the patch mechanisms to get  to where they were before. For models and operations where there were simple renames, `@projectedName` can be used to help with SDK generation. These customizations should be added in client.tsp.

catalinaperalta reviewed (commented):
Did a quick pass through, I think there are some things to improve in the TypeSpec so that we can have less handwritten code, but happy to get more context on this :)

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_operations.py:
```diff
@@ -0,0 +1,585 @@
+    @distributed_trace
+    def get_schema_by_version(
+        self, group_name: str, name: str, schema_version: int, **kwargs: Any
+    ) -> bytes:
```
No model here. We return a binary schema as per the original swagger too: https://github.com/Azure/azure-rest-api-specs/blob/116a843340e8b7127563997e7b280596d6c97654/specification/schemaregistry/data-plane/Microsoft.EventHub/stable/2023-07-01/schemaregistry.json#L146

(which Timothee said translates to : type: string, format: binary)

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_operations.py:
```diff
@@ -0,0 +1,585 @@
+        return ItemPaged(get_next, extract_data)
+
+    @distributed_trace
+    def get_schema_by_id(self, id: str, **kwargs: Any) -> bytes:
```
no model, just bytes

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_operations.py:
```diff
@@ -0,0 +1,585 @@
+        :param name: Name of schema. Required.
+        :type name: str
+        :param schema_version: Version number of specific schema. Required.
+        :type schema_version: int
```
yeah, schema version is only an integer: https://github.com/Azure/azure-rest-api-specs/blob/116a843340e8b7127563997e7b280596d6c97654/specification/schemaregistry/data-plane/Microsoft.EventHub/stable/2023-07-01/schemaregistry.json#L537

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_patch.py:
```diff
@@ -0,0 +1,294 @@
+        name: str,
+        schema_content: IO,
+        **kwargs: Any
+    ) -> None:
```
I copied this operation from code generated from the swagger (so previously generated code), as per Izzy's recommendation. The two operations that were implemented here were not generated b/c they both support multiple content types in the request, and that's not something that's been added to the generation yet.

I'm not quite sure about the typing here. But I believe, after talking to Izzy, that this was None b/c it doesn't return anything by default. It only returns an object when we pass in a response hook/cls. I tried to find a way around this before, but I don't think there is one right now. If we wanted the return to be not None/Optional[...], it would have to return the same return type of the callback passed to `cls`.

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,724 @@
+    response_headers: Mapping[str, Union[str, int]],
+) -> Tuple[Union["HttpResponse", "AsyncHttpResponse"], Dict[str, Union[int, str]]]:
+    properties_dict = _parse_schema_properties_dict(response_headers)
+    # TODO: content type is not being added to the response headers b/c of
```
I have an issue created [here](https://github.com/Azure/autorest.python/issues/2122) for this, and will add it to the comment!

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_operations.py:
```diff
@@ -0,0 +1,515 @@
+
+
+    @distributed_trace
+    def get_schema_by_version(
```
This is the only one not marked as "private" with the "_", is that on purpose?

swathipil replied:
forgot to make this op internal in the client.tsp - updating!

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/setup.py:
```diff
@@ -13,65 +13,61 @@
 # Change the PACKAGE_NAME only to change folder and different name
 PACKAGE_NAME = "azure-schemaregistry"
-PACKAGE_PPRINT_NAME = "Schema Registry"
+PACKAGE_PPRINT_NAME = "Azure Schema Registry"
```
you can add a `package-pprint-name` property in tspconfig.yaml for the python emitter to configure this value if you need any changes

swathipil replied:
for now, we're not letting the emitter overwrite the existing setup.py, b/c we have some extras_require that we don't want overwritten. But will update this in the setup.py!

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/tsp-location.yaml:
```diff
@@ -0,0 +1,4 @@
+cleanup: false
+commit: 8c473d0d5890554bc6a4b2101b7dc26bfa734b41
```
When you merge the tsp into main, you should update this commit to point to the merged commit

swathipil replied:
will do!

catalinaperalta reviewed (commented):
This is looking great, just a small comment on the operations

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_operations/_patch.py:
```diff
@@ -0,0 +1,304 @@
+_SERIALIZER = Serializer()
+
+
+def build_schema_registry_get_schema_id_by_content_request( # pylint: disable=name-too-long
```
Just curious what needed to change in these methods?

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,708 @@
+        return SchemaProperties(**properties)
+
+
+#class SchemaProperties(object):
```
Clean up?

swathipil replied:
uncommented

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_version.py:
```diff
@@ -1,27 +1,9 @@
 # --------------------------------------------------------------------------
 
-VERSION = "1.3.0b4"
+VERSION = "1.3.0b3"
```
Looks like an old version was generated.

swathipil replied:
will update!

swathipil commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_client.py:
```diff
@@ -12,42 +12,48 @@
     from azure.core.credentials import TokenCredential
 
+class SchemaRegistryClient(SchemaRegistryClientOperationsMixin):  # pylint: disable=client-accepts-api-version-keyword
+    """SchemaRegistryClient.
```
updated

swathipil commented:
/azp run python - schemaregistry - tests

swathipil commented:
/azp run python - schemaregistry - ci

swathipil commented:
/azp run python - schemaregistry - ci

swathipil commented:
/azp run python - schemaregistry - ci

swathipil commented:
/azp run python - schemaregistry - ci

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_patch.py:
```diff
@@ -0,0 +1,694 @@
+    :raises: Exception if content is invalid against provided schema.
+    """
+
+    def __call__(self, schema: Mapping[str, Any], content: Mapping[str, Any]) -> None:
```
Where are we doing the validation?

swathipil replied:
We call the callback that follows this protocol in JsonSchemaEncoder.encode/decode here: https://github.com/Azure/azure-sdk-for-python/blob/5513d6981f51196024fe2edaa9470712e4f35d65/sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/encoder/jsonencoder/_utils.py#L103.

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/models/_patch.py:
```diff
@@ -8,6 +8,23 @@
+from azure.core import CaseInsensitiveEnumMeta
+
+
+class SchemaFormat(str, Enum, metaclass=CaseInsensitiveEnumMeta):
```
Seems like we need to add this to the `__all__`?

swathipil replied:
This is currently public under the azure.schemaregistry namespace, so I didn't add it to all here, since that would move it to azure.schemaregistry.models.

catalinaperalta commented on sdk/schemaregistry/azure-schemaregistry/tsp-location.yaml:
```diff
@@ -0,0 +1,4 @@
+cleanup: false
+commit: 44cbf51b9eee8a2dbf55ced61a61032aaed0148b
```
Just make sure this is pointing to the latest commit for the tsp files.

catalinaperalta reviewed (approved):
Overall lgtm, just take a look at whether you need to expose SchemaFormat

swathipil commented:
/azp run python - schemaregistry - tests

swathipil commented:
/azp run python - schemaregistry - tests

kashifkhan commented on sdk/schemaregistry/azure-schemaregistry/setup.py:
```diff
@@ -13,65 +13,61 @@
-        'msrest>=0.6.21',
-        'azure-core>=1.24.0'
-    ],
+    install_requires=["msrest>=0.6.21", "azure-core<2.0.0,>=1.28.0"],
```
is msrest still required?

swathipil replied:
oh nice catch! it shouldn't be here at all

swathipil commented:
/azp run python - schemaregistry - tests

swathipil commented:
/azp run python - schemaregistry - ci

swathipil commented:
/azp run python - schemaregistry - tests
